### PR TITLE
UI: Breaking `ControlledTree` changes

### PR DESCRIPTION
**Pull request by grigasp**

### CODE REVIEW

**Review by pmconne — APPROVED** [review]

File: ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx [hunk 1]
```diff
@@ -26,12 +26,12 @@ import { RenderedItemsRange, TreeRenderer, TreeRendererProps } from "./TreeRende
...
-  /** Flat list of nodes to be rendered in tree. */
-  visibleNodes: VisibleTreeNodes;
+  /** Model of the tree to display. */
+  model: TreeModel;
   /** Node loader used to load root nodes and placeholder nodes. */
   nodeLoader: ITreeNodeLoader;
   /** Tree events handler. */
-  treeEvents: TreeEvents;
+  eventsHandler: TreeEvents;
   /** Mode of nodes' selection in tree. */
```
Comment (roluk) [inline ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx:35]: There is nothing wrong with the name `eventsHandler` per se, but it is more usual to use `eventHandler`. There are already some places that pass this prop like this: `eventsHandler={eventHandler}`.

**Review by roluk — CHANGES_REQUESTED** [review]

**Review by bsteinbk — APPROVED** [review]

File: presentation/components/src/test/tree/controlled/TreeHooks.test.ts [hunk 2]
```diff
@@ -33,6 +36,10 @@ describe("usePresentationNodeLoader", () => {
     pagingSize: 5,
   };
 
+  before(async () => {
+    await UiComponents.initialize(new I18N());
+  });
+
```
Comment (roluk) [inline presentation/components/src/test/tree/controlled/TreeHooks.test.ts:42]: This mutates global test environment. Either undo the changes after relevant tests complete, or perform this in the global test setup.

**Review by roluk — CHANGES_REQUESTED** [review]

**Review by mdastous-bentley — APPROVED** [review]

File: ui/framework/src/ui-framework/imodel-components/category-tree/CategoriesTree.tsx [hunk 3]
```diff
@@ -111,29 +113,24 @@ export function CategoryTree(props: CategoryTreeProps) {
   }, []);
 
   return (
-    <div className="ui-fw-categories-tree">
+    <div className="ui-fw-categories-tree" style={{ width: props.width, height: props.height }}>
       <ControlledTree
         nodeLoader={filteredNodeLoader}
```
Comment (roluk) [inline ui/framework/src/ui-framework/imodel-components/category-tree/CategoriesTree.tsx:118]: What happens when we do not set explicit dimensions on this `div`?

**Review by roluk — CHANGES_REQUESTED** [review]

File: docs/changehistory/NextVersion.md [hunk 4]
```diff
@@ -334,6 +347,80 @@ the widget in a floating container. The property `defaultFloatingPosition` may a
 
 The method `getFloatingWidgetContainerIds()` has been added to FrontstageDef to retrieve the Ids for all floating widget containers for the active frontstage as specified by the `frontstageDef`. These ids can be used to query the size of the floating container via `frontstageDef.getFloatingWidgetContainerBounds`. The method `frontstageDef.setFloatingWidgetContainerBounds` can then be used to set the size and position of a floating widget container.
 
+#### `ControlledTree` API Changes
+
+There were some breaking API changes related to `ControlledTree` component:
```
Comment (roluk) [inline docs/changehistory/NextVersion.md:352]:
```suggestion
`ControlledTree` component has received the following breaking changes:
```

File: docs/changehistory/NextVersion.md [hunk 5]
```diff
@@ -334,6 +347,80 @@ the widget in a floating container. The property `defaultFloatingPosition` may a
...
+
+* The component now takes `TreeModel` rather than `VisibleTreeNodes` as a prop to avoid requiring consumers to manage `VisibleTreeNodes` object. As a result, the `useVisibleTreeNodes` hook was replaced with `useTreeModel` hook. Typical mitigation:
```
Comment (roluk) [inline docs/changehistory/NextVersion.md:354]: Instead of `Typical mitigation` let's try `Typical fix` or `Typical migration`

File: docs/changehistory/NextVersion.md [hunk 6]
```diff
@@ -334,6 +347,80 @@ the widget in a floating container. The property `defaultFloatingPosition` may a
...
+
+  **Before:**
+
+  ```tsx
+  const visibleNodes = useVisibleTreeNodes(modelSource);
+  return (
+    <ControlledTree {...otherProps} visibleNodes={visibleNodes} />
+  );
+  ```
```
Comment (roluk) [inline docs/changehistory/NextVersion.md:363]: The code fragment could be shorter:
```tsx
const visibleNodes = useVisibleTreeNodes(modelSource);
return <ControlledTree visibleNodes={visibleNodes} {...otherProps} />
```

Also, maybe put the relevant prop before the rest.

File: docs/changehistory/NextVersion.md [hunk 7]
```diff
@@ -334,6 +347,80 @@ the widget in a floating container. The property `defaultFloatingPosition` may a
...
+
+  **After:**
+
+  ```tsx
+  return (
+    <ControlledTree {...otherProps} eventsHandler={eventsHandler} />
+  );
+  ```
+
+* Made the props `width` and `height` required. Previously they were optional and forced us to use non-optimal approach when not provided. Now it's up to the consumer to tell the size of the component. Typical mitigation:
```
Comment (roluk) [inline docs/changehistory/NextVersion.md:392]: Developers almost certainly aren't aware that `width` and `height` properties even existed as they only have been added recently. Also, we probably shouldn't try to sell people on removal of a feature when it was flawed from the start.

```suggestion
* `width` and `height` properties are now required. Typical fix:
```

File: docs/changehistory/NextVersion.md [hunk 8]
```diff
@@ -334,6 +347,80 @@ the widget in a floating container. The property `defaultFloatingPosition` may a
...
+
+  ```tsx
+  const width = 100;
+  const height = 100;
+  return (
+    <ControlledTree {...props} width={width} height={height} />
+  );
+  ```
+
+  Or, the size may be determined dynamically. The below example uses an `AutoSizer` component from `react-virtualized-auto-sizer` package for that:
```
Comment (roluk) [inline docs/changehistory/NextVersion.md:412]: We should direct people to use [`ResizeObserver`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) instead.

**Review by roluk — CHANGES_REQUESTED** [review]

File: ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx (hunk 1 above)
Comment (grigasp) [reply ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx:35]: There are also some others where the prop is assigned like `eventsHandler={eventsHandler}`... I personally like it in plural better, because the handler may handle multiple events, not just one.

File: docs/changehistory/NextVersion.md (hunk 7 above)
Comment (grigasp) [reply docs/changehistory/NextVersion.md:392]: I don't see a problem with us explaining why we removed a feature (flawed).

File: docs/changehistory/NextVersion.md (hunk 8 above)
Comment (grigasp) [reply docs/changehistory/NextVersion.md:412]: I don't mind us referencing `ResizeObserver` as a possible solution, but I don't want to give an example with it, because it would be much larger. What do you think about rephrasing the comment like:
```markdown
Or, the size may be determined dynamically using [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) or some third-party library. The below example uses an `AutoSizer` component from `react-virtualized-auto-sizer` package for that:
```
?

File: presentation/components/src/test/tree/controlled/TreeHooks.test.ts (hunk 2 above)
Comment (grigasp) [reply presentation/components/src/test/tree/controlled/TreeHooks.test.ts:42]: Fixed, thanks

File: extensions/map-layers/src/ui/widget/SubLayersTree.tsx [hunk 9]
```diff
@@ -155,14 +155,22 @@ export function SubLayersTree(props: { mapLayer: StyleMapLayerSettings }) {
...
+              nodeHighlightingProps={nodeHighlightingProps}
+              width={width}
+              height={height}
+            />
+          )}
+        </AutoSizer>
+      </div>
     </div>
   </>;
 }
```
[the remark this answers is not in the document]
Comment (grigasp) [reply extensions/map-layers/src/ui/widget/SubLayersTree.tsx]: Changed to `react-resize-detector` which uses `ResizeObserver` underneath.

File: ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx [hunk 10]
```diff
@@ -81,26 +81,29 @@ export function ControlledTree(props: ControlledTreeProps) {
...
-    <Loader loading={loading} noData={noData} spinnerRenderer={props.spinnerRenderer} noDataRenderer={props.noDataRenderer}>
-      {props.treeRenderer ? props.treeRenderer(treeProps) : <TreeRenderer {...treeProps} />}
-    </Loader>
+    <div style={{ width: props.width, height: props.height }}>
+      <Loader loading={loading} noData={noData} spinnerRenderer={props.spinnerRenderer} noDataRenderer={props.noDataRenderer}>
+        {props.treeRenderer ? props.treeRenderer(treeProps) : <TreeRenderer {...treeProps} />}
+      </Loader>
+    </div>
   );
 }
```
[the remark this answers is not in the document]
Comment (grigasp) [reply ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx]: Cleaned up

File: ui/components/src/ui-components/tree/controlled/component/TreeRenderer.tsx [hunk 11]
```diff
@@ -178,20 +178,13 @@ export class TreeRenderer extends React.Component<TreeRendererProps> implements
 
   public override render() {
     return (
-      <ConditionalAutoSizer width={this.props.width} height={this.props.height}>
-        {(size) => <TreeRendererInner ref={this.setTreeRendererRef} {...this.props} {...size} />}
-      </ConditionalAutoSizer>
+      <TreeRendererInner ref={this.setTreeRendererRef} {...this.props} />
     );
   }
 }
```
[the remark this answers is not in the document]
Comment (grigasp) [reply ui/components/src/ui-components/tree/controlled/component/TreeRenderer.tsx]: Done, thanks.

File: ui/framework/src/ui-framework/imodel-components/category-tree/CategoriesTree.tsx (hunk 3 above)
Comment (grigasp) [reply ui/framework/src/ui-framework/imodel-components/category-tree/CategoriesTree.tsx]: Cleaned up

File: docs/changehistory/NextVersion.md (hunk 5 above)
Comment (grigasp) [reply docs/changehistory/NextVersion.md]: Changed to "migration"

File: docs/changehistory/NextVersion.md (hunk 6 above)
Comment (grigasp) [reply docs/changehistory/NextVersion.md]: Done, thanks

File: docs/changehistory/NextVersion.md (hunk 8 above)
Comment (grigasp) [reply docs/changehistory/NextVersion.md]: Changed to
```markdown
`width` and `height` props may be calculated dynamically using [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) API.
```
and removed the example - let's leave it up to consumers to decide how to come up with width and height as it's fairly straightforward to find and anything we suggest may become out of date quickly.

File: ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (roluk) [reply ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx]: Thanks

File: test-apps/ui-test-app/src/frontend/appui/widgets/NavigationTreeWidget.tsx [hunk 12]
```diff
@@ -99,14 +100,19 @@ const NavigationTree: React.FC<NavigationTreeProps> = (props: NavigationTreeProp
...
+  const treeModel = useTreeModel(modelSource);
+  const { width, height, ref } = useResizeDetector();
   return (
-    <ControlledTree
-      visibleNodes={visibleNodes}
-      nodeLoader={nodeLoader}
-      selectionMode={SelectionMode.Single}
-      treeEvents={eventHandler}
-    />
+    <div ref={ref} style={{ width: "100%", height: "100%" }}>
```
Comment (roluk) [inline test-apps/ui-test-app/src/frontend/appui/widgets/NavigationTreeWidget.tsx:106]: While `VisibilityTreeWidget` has lost `width` and `height` attributes, they remain here. Is this intentional?

**Review by roluk — CHANGES_REQUESTED** [review]

File: docs/changehistory/NextVersion.md (hunk 8 above)
Comment (roluk) [reply docs/changehistory/NextVersion.md]: Agree, thanks.

File: docs/changehistory/NextVersion.md [hunk 13]
```diff
@@ -334,6 +347,58 @@ the widget in a floating container. The property `defaultFloatingPosition` may a
...
+  return <ControlledTree treeEvents={eventsHandler} {...otherProps} />;
+  ```
+
+  **After:**
+
+  ```tsx
+  return <ControlledTree eventsHandler={eventsHandler} {...otherProps} />;
+  ```
+
+* Made the props `width` and `height` required. Previously they were optional and forced us to use non-optimal approach when not provided. Now it's up to the consumer to tell the size of the component. Typical migration:
```
Comment (roluk) [inline docs/changehistory/NextVersion.md:384]: We should probably stick to present tenses for consistency.

```suggestion
* `width` and `height` properties are now required. Previously they were optional and forced us to use non-optimal approach when not provided. Now it's up to the consumer to tell the size of the component. Typical migration:
```

**Review by roluk — CHANGES_REQUESTED** [review]

File: ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx (hunk 1 above)
Comment (roluk) [reply ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx:35]: The amount of search results for "EventHandler" and "EventsHandler" on Google differ by three orders of magnitude. Also [event loop](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop#event_loop) is a widely-known concept that deals with multiple kinds of events.

Comment (grigasp) [reply ui/components/src/ui-components/tree/controlled/component/ControlledTree.tsx:35]: > The amount of search results for "EventHandler" and "EventsHandler" on Google differ by three orders of magnitude.

Top 3 results for `EventHandler` are for a case where the handler is handling only one event.

> Also event loop is a widely-known concept that deals with multiple kinds of events.

Yes, but in case of a loop it also makes sense, because each iteration of the loop is handling one event.

Honestly, I don't care which one we choose, but I don't think it's worth spending time removing "s" which is already there.

File: test-apps/ui-test-app/src/frontend/appui/widgets/NavigationTreeWidget.tsx (hunk 12 above)
Comment (grigasp) [reply test-apps/ui-test-app/src/frontend/appui/widgets/NavigationTreeWidget.tsx:106]: Yes. `VisibilityTreeWidget` gets its style from CSS through a class. This one doesn't.

**Review by roluk — APPROVED** [review]
